### ISSUE TITLE
Feature: Collect user IP address/geolocation info and store with session info

### ISSUE THREAD

**Issue by sebsilas** [pmcharrison/psychTestR#111]
Hi Peter,

What do you think about adding a feature to store users' IP addresses and geolocation information in the session object?

I've worked on a solution on my branch: https://github.com/sebsilas/psychTestR which uses a free API  to collect the IP information (https://freeipapi.com).

Here's a test (which only prints the IP, but lots more info is collected).

```r

library(psychTestR)


make_test(
  elts = join(
    one_button_page("Hi"),
    reactive_page(function(state, ...) {

      session_info <- psychTestR::get_session_info(state, complete = FALSE)

      ui <- paste0("Your IP address is ", session_info$ip_address$ipAddress)

      one_button_page(ui)
    }),
    final_page("Bye")
  )
)
```

Note that the `session_info` object can now be more than one level deep. Is that going to be a problem? If so, we could unnest session_info$ip_address and store everything in the top level appropriately.

**Comment by pmcharrison** [comment]
Thanks @sebsilas, sounds like a good feature, I would be happy to have it!

Because it is potentially personally identifying information, I think we should make recording IP addresses opt-in behaviour (i.e. you have to explicitly enable it).

I think a nested `session_info` works fine, yes.

**Comment by sebsilas** [comment]
Sure. Do you mean for the researcher (e.g., adding "record_user_ip" parameter with default to FALSE) or for the participant (or both)?

**Comment by pmcharrison** [comment]
Just the researcher!

**Comment by sebsilas** [comment]
Ok, thought so!

I think I'll add a separate argument for collecting information about the browser too via the `window.navigator` object. You can actually collect this client side unlike the IP. This can also tell you some useful features like the number of processors on a client's computer, which we've found predict performance on some tests (but note, there are multiple causal explanations for that!)

Will submit a PR when I've finished.

**Comment by pmcharrison** [comment]
That sounds handy! Great!

**Comment by klausfrieler** [comment]
Adding user_info made elt_save_results_to_disk() fail, if user_info is NULL, as it happend to me now after updating psychTestR on the DOTS server.  More specifcally happens in results.R in lines 150-151, where adding a NULL value to the place[[index_1]] list does not increase the length and then the lables  have the wrong length then.

**Comment by sebsilas** [comment]
Hi @pmcharrison, I've pushed a fix for this and submitted a PR.

**Comment by pmcharrison** [comment]
@klausfrieler, sorry about the disruption this must have caused! Thank you for providing some insight into the problem.

@sebsilas, thanks for chasing this up so quickly! Have started a discussion on the PR.

**Comment by sebsilas** [comment]
My apologies also @klausfrieler!

**Comment by pmcharrison** [comment]
@klausfrieler we have now made a new release that hopefully solves this issue for you?

**Comment by klausfrieler** [comment]
Yup, that seems to work! Thanks!

**Comment by sebsilas** [comment]
Great. Sorry for the hassle again @klausfrieler!

**Comment by pmcharrison** [comment]
Great! Closed as resolved in PR #114.
